src/Microsoft.ML.Data/Model/ModelOperationsCatalog.cs

```diff
@@ -32,13 +32,34 @@ internal ModelOperationsCatalog(IHostEnvironment env)
         /// <param name="stream">A writeable, seekable stream to save to.</param>
         public void Save(ITransformer model, Stream stream) => model.SaveTo(_env, stream);
 
+        public void Save<TSource>(IDataLoader<TSource> model, Stream stream)
```

Comment:
>public void Save<TSource>(IDataLoader<TSource> model, Stream stream) [](start = 8, length = 68)

@TomFinley's comment from #2850: 
This I think this is not correct... we need to save both the loader (or failing that, input schema) as well as the transform model to a stream. Not two separate things.

I recall the mistake that was made with predictor models and transform models in entry-points, where we viewed these as single things saved to a stream, as opposed to having to capture the entire pipeline. This mistake was inherited by ML.NET v0.1, which led to all sorts of grief when people realized, yes, saving the loader is really freakin' important. So I think we're just making the same mistake again. A model isn't just a transform model, or a loader, or an input schema, it I think needs to capture all of them, and I think we need a change a bit more fundamental than merely adding a few methods here and there. We could have done that after v1.0. I think the problem is that the API here is just misleading.

Reply:
This is an API to save an `IDataLoader`, but it can be used to save the entire model like this:

```
var loader = ml.Model.CreateTextLoader(...);
var pipeline = loader.Append(/* define a transformer here */);
var model = pipeline.Fit(/* give it an IMultiStreamSource */);
ml.Model.Save(model, fs);
```
`model` here is a `CompositeDataLoader`. 
Perhaps I should define this method to take a `CompositeDataLoader` instead of an `IDataLoader`?




---
In reply to: [262685374](https://github.com/dotnet/machinelearning/pull/2858#discussion_r262685374) [](ancestors = 262685374)

Reply:
I will add the code above to the test.

---
In reply to: [262756744](https://github.com/dotnet/machinelearning/pull/2858#discussion_r262756744) [](ancestors = 262756744,262685374)

Reply:
The problem that I see is a bit different. I understand that it's *possible* to do the right thing, but I'd like our API to be changed in such a way that doing the right thing (saving the input schema with the model) becomes easy and convenient, and doing the wrong thing (not saving the input schema) becomes less convenient, or even impossible. (Since if an `ITransformer` was trained, surely there must have been an input schema *somewhere* in that process.)

So, what I am saying is the current scheme for saving and loading models that takes a stream and an `ITransformer` and *that's it*, should go away, since it is confusing and leading people down the right path. It's not that I think your change doesn't make the right thing possible, I'm saying that by having the old way alongside it we're going to forever have the question of, "I had this problem, I don't know my input features now," and we have to forever explain, "yeah, we have this booby trap in our code, you should have used this other method, your model is now incomplete forever." Which they will, because, on the face of it, it may not be obvious to people why merely saving an `ITransformer` is insufficient. I'd rather never get that question, and I'd do that by removing the current API and replacing it with something better, that *always* takes a schema or loader. (And on the flip side of loading, always returns a loader or schema.)